pkg/proxy/proxy.go

```diff
@@ -136,130 +145,191 @@ func (p *Proxy) Run(stopCh <-chan struct{}) (<-chan struct{}, error) {
+func (p *Proxy) serve(handler http.Handler, stopCh <-chan struct{}) (<-chan struct{}, error) {
+	// Setup proxy handlers
+	handler = p.withHandlers(handler)
+
 	// securely serve using serving config
-	waitCh, err := p.secureServingInfo.Serve(proxyHandler, time.Second*60, stopCh)
+	waitCh, err := p.secureServingInfo.Serve(handler, time.Second*60, stopCh)
 	if err != nil {
 		return nil, err
 	}
 
 	return waitCh, nil
 }
 
+func (p *Proxy) withHandlers(handler http.Handler) http.Handler {
+	// Set up proxy handlers
+	handler = p.withImpersonateRequest(handler)
+	handler = p.withAuthenticateRequest(handler)
+	return handler
+}
+
+// RoundTrip is called last and is used to manipulate the forwarded request using context.
 func (p *Proxy) RoundTrip(req *http.Request) (*http.Response, error) {
-	// Clone the request here since successfully authenticating the request
-	// deletes those auth headers
-	reqCpy := utilnet.CloneRequest(req)
+	// Here we have successfully authenticated so now need to determine whether
+	// we need use impersonation or not.
+
+	// If no impersonation then we return here without setting impersonation
+	// header but re-introduce the token we removed.
+	if context.NoImpersonation(req.Context()) {
+		token := context.BearerToken(req.Context())
+		req.Header.Add("Authorization", token)
+		return p.noAuthClientTransport.RoundTrip(req)
+	}
 
-	// auth request and handle unauthed
-	info, ok, err := p.oidcRequestAuther.AuthenticateRequest(reqCpy)
-	if err != nil {
+	// Get the impersonation headers from the context.
+	conf := context.ImpersonationConfig(req.Context())
+	if conf == nil {
+		return nil, errNoImpersonationConfig
+	}
 
-		// attempt to passthrough request if valid token
-		if p.options.TokenReview {
-			return p.tokenReview(reqCpy)
-		}
+	// Set up impersonation request.
+	rt := transport.NewImpersonatingRoundTripper(*conf, p.clientTransport)
 
-		return nil, errUnauthorized
-	}
+	// Push request through round trippers to the API server.
+	return rt.RoundTrip(req)
+}
 
-	// failed authorization
-	if !ok {
-		return nil, errUnauthorized
-	}
+// withAuthenticateRequest adds the proxy authentication handler to a chain.
+func (p *Proxy) withAuthenticateRequest(handler http.Handler) http.Handler {
+	return http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+		// Auth request and handle unauthed
+		info, ok, err := p.oidcRequestAuther.AuthenticateRequest(req)
+		if err != nil {
+			if !p.config.TokenReview {
+				p.handleError(rw, req, errUnauthorized)
+				return
+			}
+
+			// Attempt to passthrough request if valid token
+			if p.reviewToken(rw, req) {
+				// Set no impersonation headers and re-add removed headers.
+				req = req.WithContext(context.WithNoImpersonation(req.Context()))
+
+				handler.ServeHTTP(rw, req)
+				return
+			}
+
+			// Token review failed so error
+			p.handleError(rw, req, errUnauthorized)
+			return
+		}
```

Comment:
That's a lot of code in a `if err != nil` block.. can this be pulled into its own function? so `if err != nil { return doSomething() }`?

Reply:
I can try a "noAuth" type func, not sure